source/common/common/compiled_string_map.h

```diff
@@ -0,0 +1,157 @@
+      return {};
+    }
+    return table_[key.size()](key);
+  };
+  /**
+   * Construct the lookup table. This is a somewhat slow multi-pass
+   * operation - using this structure is not recommended unless the
+   * table is initialize-once, use-many.
+   * @param initial a vector of key->value pairs.
+   */
+  void compile(std::vector<KV> initial) {
+    if (initial.empty()) {
+      return;
+    }
+    size_t longest = 0;
+    for (const KV& pair : initial) {
+      longest = std::max(pair.first.size(), longest);
+    }
+    table_.resize(longest + 1);
+    std::sort(initial.begin(), initial.end(),
+              [](const KV& a, const KV& b) { return a.first.size() < b.first.size(); });
+    auto it = initial.begin();
+    for (size_t i = 0; i <= longest; i++) {
+      auto start = it;
+      while (it != initial.end() && it->first.size() == i) {
+        it++;
+      }
+      if (it != start) {
+        std::vector<KV> node_contents;
+        node_contents.reserve(it - start);
+        std::copy(start, it, std::back_inserter(node_contents));
+        table_[i] = createEqualLengthNode(node_contents);
+      }
+    }
+  }
+
+private:
+  static FindFn createEqualLengthNode(std::vector<KV> node_contents) {
+    if (node_contents.size() == 1) {
+      return [pair = node_contents[0]](const absl::string_view& key) -> Value {
+        if (key != pair.first) {
+          return {};
+        }
+        return pair.second;
+      };
+    }
+    struct IndexSplitInfo {
+      uint8_t index, min, max, count;
+    } best{0, 0, 0, 0};
+    for (size_t i = 0; i < node_contents[0].first.size(); i++) {
+      std::array<bool, 256> hits{};
+      IndexSplitInfo info{static_cast<uint8_t>(i), 255, 0, 0};
+      for (size_t j = 0; j < node_contents.size(); j++) {
+        uint8_t v = node_contents[j].first[i];
+        if (!hits[v]) {
+          hits[v] = true;
+          info.count++;
+          info.min = std::min(v, info.min);
+          info.max = std::max(v, info.max);
+        }
+      }
+      if (info.count > best.count) {
+        best = info;
+      }
+    }
+    std::vector<FindFn> nodes;
+    nodes.resize(best.max - best.min + 1);
+    std::sort(node_contents.begin(), node_contents.end(), [&best](const KV& a, const KV& b) {
+      return a.first[best.index] < b.first[best.index];
+    });
+    auto it = node_contents.begin();
+    for (int i = best.min; i <= best.max; i++) {
+      auto start = it;
+      while (it != node_contents.end() && it->first[best.index] == i) {
+        it++;
+      }
+      if (it != start) {
+        // Optimization was tried here, std::array<KV, 256> rather than
+        // a smaller-range vector with bounds, to keep locality and reduce
+        // comparisons. It didn't help.
+        std::vector<KV> next_contents;
+        next_contents.reserve(it - start);
+        std::copy(start, it, std::back_inserter(next_contents));
+        nodes[i - best.min] = createEqualLengthNode(next_contents);
```

Comment:
If I was doing it in C it would be *almost* the same - I don't know how much overhead there is in std::function but for the C structure of this I'd be using something like a struct containing a function pointer and a union of the captured data that each of the node-types takes, and that function pointer would take a pointer to that same structure... which seems like it's probably at least **about** the same as what a lambda is doing.

It might be possible to optimize further with more manual constructs, but I think a 60% speedup is probably good enough, and it could have another pass later if someone wants to try to do better.